packages-exp/app-check-exp/src/internal-api.test.ts

```diff
@@ -0,0 +1,443 @@
+      );
+      const listener1 = (): void => {
+        throw new Error();
+      };
+      const listener2 = spy();
+
+      addTokenListener(app, fakePlatformLoggingProvider, listener1);
+      addTokenListener(app, fakePlatformLoggingProvider, listener2);
+
+      await getToken(app, fakePlatformLoggingProvider);
+
+      expect(listener2).to.be.calledWith({
+        token: fakeRecaptchaAppCheckToken.token
+      });
+    });
+
+    it('loads persisted token to memory and returns it', async () => {
+      const clock = useFakeTimers();
+      initializeAppCheck(app, {
+        provider: new ReCaptchaV3Provider(FAKE_SITE_KEY)
+      });
+
+      stub(storage, 'readTokenFromStorage').returns(
+        Promise.resolve(fakeCachedAppCheckToken)
+      );
+
+      const clientStub = stub(client, 'exchangeToken');
+
+      expect(getState(app).token).to.equal(undefined);
+      expect(await getToken(app, fakePlatformLoggingProvider)).to.deep.equal({
+        token: fakeCachedAppCheckToken.token
+      });
+      expect(getState(app).token).to.equal(fakeCachedAppCheckToken);
+      expect(clientStub).has.not.been.called;
+
+      clock.restore();
+    });
+
+    it('persists token to storage', async () => {
+      initializeAppCheck(app, {
+        provider: new ReCaptchaV3Provider(FAKE_SITE_KEY)
+      });
+
+      stub(storage, 'readTokenFromStorage').returns(Promise.resolve(undefined));
+      stub(reCAPTCHA, 'getToken').returns(Promise.resolve(fakeRecaptchaToken));
+      stub(client, 'exchangeToken').returns(
+        Promise.resolve(fakeRecaptchaAppCheckToken)
+      );
+      const storageWriteStub = stub(storage, 'writeTokenToStorage');
+      const result = await getToken(app, fakePlatformLoggingProvider);
+      expect(result).to.deep.equal({ token: fakeRecaptchaAppCheckToken.token });
+      expect(storageWriteStub).has.been.calledWith(
+        app,
+        fakeRecaptchaAppCheckToken
+      );
+    });
+
+    it('returns the valid token in memory without making network request', async () => {
+      const clock = useFakeTimers();
+      initializeAppCheck(app, {
+        provider: new ReCaptchaV3Provider(FAKE_SITE_KEY)
+      });
+      setState(app, { ...getState(app), token: fakeRecaptchaAppCheckToken });
+
+      const clientStub = stub(client, 'exchangeToken');
+      expect(await getToken(app, fakePlatformLoggingProvider)).to.deep.equal({
+        token: fakeRecaptchaAppCheckToken.token
+      });
+      expect(clientStub).to.not.have.been.called;
+
+      clock.restore();
+    });
+
+    it('force to get new token when forceRefresh is true', async () => {
+      initializeAppCheck(app, {
+        provider: new ReCaptchaV3Provider(FAKE_SITE_KEY)
+      });
+      setState(app, { ...getState(app), token: fakeRecaptchaAppCheckToken });
+
+      stub(reCAPTCHA, 'getToken').returns(Promise.resolve(fakeRecaptchaToken));
+      stub(client, 'exchangeToken').returns(
+        Promise.resolve(fakeRecaptchaAppCheckToken)
```

Comment:
We can't tell whether we got a new token or not if it's the same one as the cached one. Can we?

Reply:
Changed the contents of the stubbed return promise.